examples/circuit-relaying/src/helpers.js

```diff
@@ -53,22 +55,64 @@ module.exports = (ipfs, peersSet) => {
     const tags = Array.from(peersSet).map((p) => {
       return `<li>${p}</li>`
     })
-    $peers.innerHTML = tags
+    $peers.innerHTML = tags.join('')
   }
 
-  const updateAddrs = (addrs) => {
-    $addrs.innerHTML = `
-        <div>
-            <ul>
-                ${addrs.map((addr) => `<li>${addr.toString()}</li>`)}
-            <ul>
-        </div>`
+  const updateSwarmPeers = async (ipfs) => {
+    const peers = await ipfs.swarm.peers()
+
+    $pAddrs.innerHTML = peers.map(peer => `<li>${peer.peer}</li>`).join('')
+  }
+
+  const updateAddrs = async (ipfs) => {
+    const info = await ipfs.id()
+
+    // see which peers support the circuit relay protocol
+    const relayAddrs = []
+    const connections = ipfs.libp2p.connections
+    const peers = await ipfs.swarm.peers()
+
+    for (let i = 0; i < peers.length; i++) {
+      const {
+        peer: peerId
+      } = peers[i]
+
+      const cons = connections.get(peerId)
+
+      for (let j = 0; j < cons.length; j++) {
+        const con = cons[j]
+        const remoteAddr = con.remoteAddr.toString()
+
+        for (let k = 0; k < 5; k++) {
+          // protocols is undefined until the connection is negotiated so try a few times
+          const protocols = ipfs.libp2p.peerStore.protoBook.get(con.remotePeer)
```

Comment:
This was a bit weird, we've called `ipfs.swarm.connect` at this point, so we're connected, but the protobook returns undefined for a little while after connecting so we can't actually use the connection yet.

I guess the connection has not been fully negotiated - might be be worth having `libp2p.dial` only return after the connection is ready to be used?

Reply:
You can listen for protocol changes on a peer now, https://github.com/libp2p/js-libp2p/blob/master/doc/API.md#known-protocols-for-a-peer-change. This prevents the need to iterate over your connections as you can just wait for the relay protocol, but this is going to trigger for every peer (STOP is on by default). Currently there's no way to know if the peer is a relay without asking it if it supports `HOP`, which is currently done by trying to dial through it, AutoRelay needs to be added to js-libp2p for this, or at least the ability to check the HOP status.

```js
ipfs.libp2p.peerStore.on('change:protocols', ({ peerId, protocols}) => {})
```

The connection is ready, but identify is being run in parallel as to not block. There's nothing prohibiting you from using the connection once dial has been completed, you just might not know their supported protocols yet.